src/provider/model/ClientRegistry.ts

```diff
@@ -0,0 +1,109 @@
+    private _apiHandler!: APIHandler<APITopic, Events>;
+
+    private _activeClients: Identity[] = [];
+    public readonly onClientHandshake: Signal<[Identity]> = new Signal();
+
+    constructor(@inject(Inject.API_HANDLER) apiHandler: APIHandler<APITopic, Events>) {
+        this._apiHandler = apiHandler;
+        this._apiHandler.onConnection.add(this.onClientConnection, this);
+        this._apiHandler.onDisconnection.add(this.onClientDisconnection, this);
+    }
+
+    /**
+     * Launches the app with specified identity if the start-up information about it was stored previously.
+     *
+     * @param app Identity of target client.
+     */
+    public tryLaunchApplication(app: Identity): void {
+        fin.Application.wrapSync(app).isRunning().then(async running => {
+            if (!running) {
+                try {
+                    const collection = Injector.get<'DATABASE'>(Inject.DATABASE).get(CollectionMap.CLIENTS);
+                    const initData = await collection.get(app.uuid);
+                    if (initData && initData.data) {
+                        if (typeof initData.data === 'string') {
+                            await fin.Application.startFromManifest(initData.data);
+                        } else {
+                            await fin.Application.start(initData.data as ApplicationOption);
+                        }
+                    } else {
+                        throw new Error('Could not find application initialization data for the application with uuid ' + app.uuid + ' in the database.');
+                    }
+                } catch (error) {
+                    this.logError(error);
+                }
+            }
+        });
+    }
+
+    /**
+     * Decides weather an app is running at the moment.
+     *
+     * @param app Identity of the app.
+     */
+    public isAppActive(app: Identity): boolean {
+        return this._activeClients.some(client => client.uuid === app.uuid);
+    }
+
+    public async onAddEventListener(eventType: string, sender: Identity): Promise<void> {
+        if (eventType === 'notification-action') {
+            this._activeClients.push(sender);
+            this.onClientHandshake.emit(sender);
+        }
+    }
+
+    public async onRemoveEventListener(eventType: string, sender: Identity): Promise<void> {
+        if (eventType === 'notification-action') {
+            this._activeClients = this._activeClients.filter(client => client.uuid !== sender.uuid || client.name !== sender.name);
+        }
+    }
+
+    private onClientConnection(app: Identity): void {
+        fin.Application.wrapSync(app).getInfo().then(async info => {
+            try {
+                const collection = Injector.get<'DATABASE'>(Inject.DATABASE).get(CollectionMap.CLIENTS);
+                await collection.upsert({
+                    id: app.uuid,
+                    data: info.parentUuid ? info.initialOptions as ApplicationOption : info.manifestUrl
```

Comment:
Suggest use a discriminated union to make things more explicit, with `type` being `manifest` or `programmatic`. I think we'll also need to store the `parentUuid` for programmatic apps, and use that when restarting them. Might be worth looking at `WorkapceApp` in layouts, and how `parentUuid` gets used - layouts does some jazz to fake that the app has been started by it's original parent

Reply:
done the discriminated union bit. let's discuss the feasibility and motivation of storing parentUuid